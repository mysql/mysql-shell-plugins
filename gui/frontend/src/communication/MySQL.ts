/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

// Have to disable the spell checker and naming convention rule as we list a number of values here that cannot be
// changed to conform to our rules.

/* spell-checker: disable */
/* eslint-disable @typescript-eslint/naming-convention */

export enum MySQLConnectionScheme {
    MySQL = "mysql",
    MySQLx = "mysqlx",
}

export enum MySQLSslMode {
    Disabled = "DISABLED",
    Preferred = "PREFERRED",
    Required = "REQUIRED",
    VerifyCA = "VERIFY_CA",
    VerifyIdentity = "VERIFY_IDENTITY",
}

export enum MySQLAuthMethod {
    Auto = "AUTO",
    FromCapabilities = "FROM_CAPABILITIES",
    Fallback = "FALLBACK",
    MySQL41 = "MYSQL41",
    Plain = "PLAIN",
    SHA256 = "SHA256_MEMORY",
}

export enum MySQLSqlMode {
    Ansi = "ANSI",
    Traditional = "TRADITIONAL",
    AllowInvalidDates = "ALLOW_INVALID_DATES",
    AnsiQuotes = "ANSI_QUOTES",
    ErrorForDivisionByZero = "ERROR_FOR_DIVISION_BY_ZERO",
    HighNotPrecendence = "HIGH_NOT_PRECEDENCE",
    IgnoreSpace = "IGNORE_SPACE",
    NoAutoValueOnZero = "NO_AUTO_VALUE_ON_ZERO",
    NoUnsignedSubtraction = "NO_UNSIGNED_SUBTRACTION",
    NoZeroDate = "NO_ZERO_DATE",
    NoZeroInDate = "NO_ZERO_IN_DATE",
    OnlyFullGroupBy = "ONLY_FULL_GROUP_BY",
    PadCharToFullLength = "PAD_CHAR_TO_FULL_LENGTH",
    PipesAsConcat = "PIPES_AS_CONCAT",
    RealAsFloat = "REAL_AS_FLOAT",
    StrictAllTables = "STRICT_ALL_TABLES",
    StrictTransTables = "STRICT_TRANS_TABLES",
    TimeTruncateFractial = "TIME_TRUNCATE_FRACTIONAL",
}

export enum MySQLConnCompression {
    "REQUIRED",  // Connection will only be made when compression can be enabled.
    "PREFERRED", // Compression is negotiated (default for X protocol).
    "DISABLED",  // No compression (default for classic protocol).
}

export interface IMySQLConnectionOptions {
    // Base options:

    // The protocol to be used on the connection.
    scheme: MySQLConnectionScheme;

    // The MySQL user name to be used on the connection.
    user?: string;

    // The password to be used on the connection.
    password?: string;

    // The hostname or IP address to be used on the connection.
    host: string;

    // The port to be used in a TCP connection.
    port?: number;

    // The socket file name for unix sockets or the pipe name for Windows pipes.
    socket?: string;

    // The schema to be selected once the connection is done.
    schema?: string;

    // Main options:

    // The SSL mode to be used in the connection.
    "ssl-mode"?: MySQLSslMode;

    // The path to the X509 certificate authority file in PEM format.
    "ssl-ca"?: string;

    // Path to the directory that contains the X509 certificate authority files in PEM format.
    "ssl-capath"?: string;

    // The path to the SSL public key certificate file in PEM format.
    "ssl-cert"?: string;

    // The path to the SSL private key file in PEM format.
    "ssl-key"?: string;

    // The path to file that contains certificate revocation lists.
    "ssl-crl"?: string;

    // The path of directory that contains certificate revocation list files.
    "ssl-crlpath"?: string;

    // The list of permissible encryption ciphers for connections that use TLS protocols up through TLSv1.2.
    // Colon-separated string. Example: "HIGH:!SSLv2:!RC4:!aNULL@STRENGTH".
    "ssl-ciphersuites"?: string;

    // Comma separated list of protocols permitted for secure connections.
    "tls-version"?: string;

    // List of TLS v1.3 ciphers to use. Same format as for field sslCipher.
    "tls-ciphers"?: string;

    // Authentication method.
    "auth-method"?: MySQLAuthMethod;

    // Request public key from the server required for RSA key pair-based password exchange. Use when connecting to
    // MySQL 8.0 servers with classic MySQL sessions with SSL mode DISABLED.
    "get-server-public-key"?: boolean;

    // The path name to a file containing a client-side copy of the public key required by the server for RSA
    // key pair-based password exchange. Use when connecting to MySQL 8.0 servers with classic MySQL sessions with
    // SSL mode DISABLED.
    "server-public-key-path"?: string;

    // The connection timeout in milliseconds. If not provided a default timeout of 10 seconds will be used.
    // Specifying a value of 0 disables the connection timeout.
    "connect-timeout"?: number;

    // Enable compression in client/server protocol.
    compression?: MySQLConnCompression;

    // Use compression algorithm in server/client protocol.
    // Comma-separated list with these values: "zlib", "zstd", "lz" and/or "uncompressed".
    "compression-algorithms"?: string;

    // Use this compression level in the client/server protocol.
    //   zstd: 1-22 (default 3)
    //   zlib: 1-9 (default 3), supported only by X protocol
    //   lz4: 0-16 (default 2), supported only by X protocol.
    "compression-level"?: number;

    // List of connection attributes to be registered at the PERFORMANCE_SCHEMA connection attributes tables.
    "connection-attributes"?: { [key: string]: string };

    // Enable/disable LOAD DATA LOCAL INFILE.
    "local-infile"?: boolean;

    // an SSHURI
    "ssh"?: string;

    // passowrd to be used on the ssh connection
    "ssh-password"?: string;

    // the ssh key file password
    "ssh-identity-file"?: string;

    // the key file to be used on the ssh connection
    "ssh-identity-file-password"?: string;

    // custom path for the ssh config file
    "ssh-config-file"?: string;

    // the ssh public key file password
    "ssh-public-identity-file"?: string;

    // mds db sytem id
    "mysql-db-system-id"?: string;

    // mds profile name
    "profile-name"?: string;

    // mds bastion id
    "bastion-id"?: string;
}
