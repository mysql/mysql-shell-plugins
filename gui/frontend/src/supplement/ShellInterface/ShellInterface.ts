/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { ShellInterfaceCore } from "./ShellInterfaceCore";
import { ShellInterfaceDbConnection } from "./ShellInterfaceDbConnection";
import { ShellInterfaceModule } from "./ShellInterfaceModule";
import { ShellInterfaceUser } from "./ShellInterfaceUser";

/**
 * This class serves as central point for singleton shell interfaces like `core` or `modules`.
 * Singletons are all those shell interfaces, which don't need a session for their work.
 */
export class ShellInterface {
    private static interfaces: {
        core?: ShellInterfaceCore;
        users?: ShellInterfaceUser;
        modules?: ShellInterfaceModule;
        dbConnections?: ShellInterfaceDbConnection;
    } = {};

    public static get core(): ShellInterfaceCore {
        if (!ShellInterface.interfaces.core) {
            ShellInterface.interfaces.core = new ShellInterfaceCore();
        }

        return ShellInterface.interfaces.core;
    }

    public static get users(): ShellInterfaceUser {
        if (!ShellInterface.interfaces.users) {
            ShellInterface.interfaces.users = new ShellInterfaceUser();
        }

        return ShellInterface.interfaces.users;
    }

    public static get modules(): ShellInterfaceModule {
        if (!ShellInterface.interfaces.modules) {
            ShellInterface.interfaces.modules = new ShellInterfaceModule();
        }

        return ShellInterface.interfaces.modules;
    }

    public static get dbConnections(): ShellInterfaceDbConnection {
        if (!ShellInterface.interfaces.dbConnections) {
            ShellInterface.interfaces.dbConnections = new ShellInterfaceDbConnection();
        }

        return ShellInterface.interfaces.dbConnections;
    }
}
