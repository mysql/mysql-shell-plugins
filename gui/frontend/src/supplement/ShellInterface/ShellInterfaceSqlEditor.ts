/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { EventType, IDispatchEvent, ListenerEntry } from "../Dispatch";
import {
    ProtocolGui, ICommErrorEvent, ICommStartSessionEvent, ShellAPIGui, ICommSimpleResultEvent,
    ShellPromptResponseType, IPromptReplyBackend, MessageScheduler, IGenericResponse,
} from "../../communication";
import { webSession } from "../WebSession";
import { settings } from "../Settings/Settings";
import { ShellInterfaceDb, ShellInterfaceMds, ShellInterfaceMrs } from ".";

export class ShellInterfaceSqlEditor extends ShellInterfaceDb implements IPromptReplyBackend {

    public mds: ShellInterfaceMds = new ShellInterfaceMds();
    public mrs: ShellInterfaceMrs = new ShellInterfaceMrs();

    public get id(): string {
        return "sqlEditor";
    }

    /**
     * @returns a flag which indicates if a session was opened already.
     */
    public get hasSession(): boolean {
        return this.moduleSessionId !== undefined;
    }

    /**
     * Begins a new session for a specific SQL editor. This overrides the underlying DB session handling.
     *
     * @param id A value identifying the SQL editor.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public startSession(id: string): Promise<void> {
        return new Promise((resolve, reject) => {
            this.moduleSessionLookupId = this.id + "." + id;
            this.mrs.moduleSessionLookupId = this.moduleSessionLookupId;

            if (this.hasSession) {
                resolve();
            }

            const request = ProtocolGui.getRequestSqleditorStartSession();
            const listener = MessageScheduler.get.sendRequest(request,
                { messageClass: ShellAPIGui.GuiSqleditorStartSession });

            listener.then((event: ICommStartSessionEvent) => {
                if (event.data) {
                    webSession.setModuleSessionId(this.moduleSessionLookupId, event.data.result.moduleSessionId);

                    resolve();
                }
            }).catch((event: ICommErrorEvent) => {
                reject(event.message);
            });
        });
    }

    /**
     * Closes this editor session and all open connections.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public closeSession(): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                resolve();
            } else {
                const request = ProtocolGui.getRequestSqleditorCloseSession(id);
                MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorCloseSession })
                    .then(() => {
                        webSession.setModuleSessionId(this.moduleSessionLookupId);
                        resolve();
                    }).catch((event: ICommErrorEvent) => {
                        reject(event.message);
                    });
            }
        });
    }

    /**
     * Returns information of how the modules should be displayed in the gui.
     *
     * @returns A listener for the response
     */
    public getGuiModuleDisplayInfo(): ListenerEntry {
        const request = ProtocolGui.getRequestSqleditorGetGuiModuleDisplayInfo();

        return MessageScheduler.get.sendRequest(request, { messageClass: "getModuleDisplayInfo" });
    }

    /**
     * Returns true as this extension object holds the backend implementation of a gui module.
     *
     * @returns A listener for the response
     */
    public isGuiModuleBackend(): ListenerEntry {
        const request = ProtocolGui.getRequestSqleditorIsGuiModuleBackend();

        return MessageScheduler.get.sendRequest(request, { messageClass: "isGuiModuleBackend" });
    }

    /**
     * Opens the database connection.
     *
     * @param dbConnectionId The id of the db connection.
     *
     * @returns A listener for the response
     */
    public openConnection(dbConnectionId: number): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorOpenConnection(dbConnectionId, id);

        return MessageScheduler.get.sendRequest(request, { messageClass: "openConnection" });
    }

    /**
     * Executes the given SQL.
     *
     * @param sql The sql command to execute.
     * @param params The module session the function should operate on.
     *
     * @returns A listener for the response
     */
    public execute(sql: string, params?: string[]): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            const response: IDispatchEvent<IGenericResponse> = {
                id: "",
                eventType: EventType.ErrorResponse,
                message: "No open session",
                context: { messageClass: "execute" },
                data: { requestState: { type: "ERROR", msg: "" } },
            };

            return ListenerEntry.resolve(response);
        }

        const request = ProtocolGui.getRequestSqleditorExecute(sql, id, params,
            { rowPacketSize: settings.get("sql.rowPacketSize", 1000) });

        return MessageScheduler.get.sendRequest(request, { messageClass: "execute" });
    }

    /**
     * Reconnects the database connection.
     *
     * @returns A listener for the response
     */
    public reconnect(): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorReconnect(id);

        return MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorReconnect });
    }

    /**
     * Stops the currently running query (if there's any).
     *
     * @returns A listener for the response
     */
    public killQuery(): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestSqleditorKillQuery(id);

        return MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorKillQuery });
    }

    /**
     * Sets the auto commit mode for the current connection.
     * Note: this mode can implicitly be changed by executing certain SQL code (begin, set autocommit, rollback, etc.).
     *
     * @param value A flag indicating if the mode should be enabled or disabled.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public setAutoCommit(value: boolean): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve();
            }

            const request = ProtocolGui.getRequestSqleditorSetAutoCommit(id, value);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorSetAutoCommit })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve();
                    }
                })
                .catch((event) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Returns the current auto commit mode, if supported.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public getAutoCommit(): Promise<boolean> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve(false);
            }

            const request = ProtocolGui.getRequestSqleditorGetAutoCommit(id);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorGetAutoCommit })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve((event.data?.result as number) !== 0);
                    }
                })
                .catch((event) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Returns the current default schema, if supported.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public getCurrentSchema(): Promise<string> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve("");
            }

            const request = ProtocolGui.getRequestSqleditorGetCurrentSchema(id);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorGetCurrentSchema })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve(event.data ? event.data.result as string : "");
                    }
                }).catch((event: ICommErrorEvent) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Sets the current default schema, if supported.
     *
     * @param schema The schema to set as the default.
     *
     * @returns A promise which resolves when the operation was concluded.
     */
    public setCurrentSchema(schema: string): Promise<void> {
        return new Promise((resolve, reject) => {
            const id = this.moduleSessionId;
            if (!id) {
                return resolve();
            }

            const request = ProtocolGui.getRequestSqleditorSetCurrentSchema(id, schema);
            MessageScheduler.get.sendRequest(request, { messageClass: ShellAPIGui.GuiSqleditorSetCurrentSchema })
                .then((event: ICommSimpleResultEvent) => {
                    if (event.eventType === EventType.FinalResponse) {
                        resolve();
                    }
                }).catch((event: ICommErrorEvent) => {
                    reject(event.message);
                });
        });
    }

    /**
     * Sends a reply from the user back to the backend (e.g. passwords, choices etc.).
     *
     * @param requestId The same request ID that was used to request input from the user.
     * @param type Indicates if the user accepted the request or cancelled it.
     * @param reply The reply from the user.
     *
     * @returns A listener for the response.
     */
    public sendReply(requestId: string, type: ShellPromptResponseType, reply: string): ListenerEntry {
        const id = this.moduleSessionId;
        if (!id) {
            return ListenerEntry.resolve();
        }

        const request = ProtocolGui.getRequestPromptReply(requestId, type, reply, id);

        return MessageScheduler.get.sendRequest(request, { messageClass: "sendReply" });
    }
}
