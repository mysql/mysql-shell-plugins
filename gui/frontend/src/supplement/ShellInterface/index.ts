/*
 * Copyright (c) 2020, 2023, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General License: , version .=> 0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General License: , version 2.0, for mor => details.
 *
 * You should have received a copy of the GNU Genera => License:
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { IDictionary } from "../../app-logic/Types";
import { IMySQLConnectionOptions } from "../../communication/MySQL";
import { ISqliteConnectionOptions } from "../../communication/Sqlite";

/** Database types we can handle. */
export enum DBType {
    MySQL = "MySQL",
    Sqlite = "Sqlite",
}

type IShellConnectionOptions = IMySQLConnectionOptions | ISqliteConnectionOptions | IDictionary;

export interface IConnectionDetails {
    // A running number in the backend DB, where connections are stored.
    id: number;

    dbType: DBType;
    caption: string;
    description: string;
    options: IShellConnectionOptions;
    useSSH?: boolean;
    useMDS?: boolean;

    version?: number;     // The version of the server. Valid not before the connection is open.
    sqlMode?: string;     // Ditto, if the server supports sql modes (MySQL).

    hideSystemSchemas?: boolean;
}

// All available shell (backend) interfaces to access the functionality of the backend.

export interface IShellSessionDetails {
    sessionId: number;    // A running number to identify open shell sessions in the UI.
    caption?: string;     // The text of the shell console tab entry and the title used in shell session tiles.
    description?: string; // A description text used in shell session tiles.

    version?: number;     // The version of the server. Valid not before the session is open.
    sqlMode?: string;     // Ditto, if the server supports sql modes (MySQL).

    dbConnectionId?: number; // The id of the database connection to open
}

export interface IBackendInformation {
    architecture: string;
    major: number;
    minor: number;
    patch: number;
    platform: string;
    serverDistribution: string;
    serverMajor: number;
    serverMinor: number;
    serverPatch: number;
}
