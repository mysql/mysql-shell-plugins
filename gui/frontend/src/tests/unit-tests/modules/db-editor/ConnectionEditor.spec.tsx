/*
 * Copyright (c) 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";

import { screen, waitFor } from "@testing-library/preact";
import { IMySQLConnectionOptions, MySQLConnectionScheme } from "../../../../communication/MySQL.js";
import { ConnectionEditor } from "../../../../modules/db-editor/ConnectionEditor.js";
import { ShellInterface } from "../../../../supplement/ShellInterface/ShellInterface.js";
import { ShellInterfaceSqlEditor } from "../../../../supplement/ShellInterface/ShellInterfaceSqlEditor.js";
import { DBType, IConnectionDetails } from "../../../../supplement/ShellInterface/index.js";
import { webSession } from "../../../../supplement/WebSession.js";
import { MySQLShellLauncher } from "../../../../utilities/MySQLShellLauncher.js";
import {
    DialogHelper, changeInputValue, getDbCredentials, nextRunLoop, setupShellForTests,
} from "../../test-helpers.js";

describe("ConnectionEditor tests", (): void => {
    let launcher: MySQLShellLauncher;
    let backend: ShellInterfaceSqlEditor;
    let connID: number;
    let dialogHelper: DialogHelper;

    const credentials = getDbCredentials();

    const options: IMySQLConnectionOptions = {
        scheme: MySQLConnectionScheme.MySQL,
        user: credentials.userName,
        password: credentials.password,
        host: credentials.host,
        port: credentials.port,
    };

    const testMySQLConnection: IConnectionDetails = {
        id: -1,
        dbType: DBType.MySQL,
        caption: "DBEditorModule Test Connection 1",
        description: "DBEditorModule Test MyQSL Connection",
        options,
    };

    beforeAll(async () => {
        backend = new ShellInterfaceSqlEditor();

        launcher = await setupShellForTests(false, true, "DEBUG3");
        testMySQLConnection.id = await ShellInterface.dbConnections.addDbConnection(webSession.currentProfileId,
            testMySQLConnection, "unit-tests") ?? -1;
        expect(testMySQLConnection.id).toBeGreaterThan(-1);
        connID = testMySQLConnection.id;

        await backend.startSession("dbEditorModuleTests");
        await backend.openConnection(testMySQLConnection.id);
        dialogHelper = new DialogHelper("connectionEditor");
    });

    afterAll(async () => {
        await backend.closeSession();
        await ShellInterface.dbConnections.removeDbConnection(webSession.currentProfileId, connID);
        await launcher.exitProcess();
    });

    it("Test ConnectionEditor match snapshot", () => {
        const component = mount<ConnectionEditor>(
            <ConnectionEditor connections={[]} onAddConnection={() => {}} onUpdateConnection={() => {}} />,
        );

        expect(component).toMatchSnapshot();
        component.unmount();
    });

    it("Test ConnectionEditor set connection timeout", async () => {
        const component = mount<ConnectionEditor>(
            <ConnectionEditor
                connections={[testMySQLConnection]}
                onAddConnection={() => { }}
                onUpdateConnection={() => { }} />,
        );

        let portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);

        await component.instance().show("MySQL", false, testMySQLConnection);

        await waitFor(() => {
            expect(screen.getByText("Advanced")).toBeDefined();
        });

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(1);

        const advanced = screen.getByText("Advanced");
        advanced.click();
        await waitFor(() => {
            expect(screen.getByText("Connection Timeout")).toBeDefined();
        });

        const timeout = screen.getByText("Connection Timeout");
        changeInputValue(timeout, "5");
        await nextRunLoop();

        dialogHelper.verifyErrors();
        await dialogHelper.clickOk();

        await nextRunLoop();
        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);

        component.unmount();
    });
});
