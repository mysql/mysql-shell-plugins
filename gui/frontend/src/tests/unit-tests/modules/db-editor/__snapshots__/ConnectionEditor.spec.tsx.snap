// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionEditor tests Test ConnectionEditor match snapshot 1`] = `
<ConnectionEditor
  connections={[]}
  onAddConnection={[Function]}
  onUpdateConnection={[Function]}
>
  <ValueEditDialog
    id="connectionEditor"
    onClose={[Function]}
    onSelectTab={[Function]}
    onValidate={[Function]}
  >
    <Dialog
      actions={
        {
          "begin": [],
          "end": [
            <Button
              caption="OK"
              disabled={false}
              id="ok"
              onClick={[Function]}
            />,
            <Button
              caption="Cancel"
              id="cancel"
              onClick={[Function]}
            />,
          ],
        }
      }
      caption={
        <g>
          <Icon
            disabled={false}
            src={391}
          />
          <Label />
        </g>
      }
      className="msg valueEditDialog"
      container={
        <body
          theme="dark"
        />
      }
      content={
        <Grid
          columnGap={16}
          columns={8}
          id="contentGrid"
          rowGap={16}
        />
      }
      id="connectionEditor"
      onClose={[Function]}
      onSelectTab={[Function]}
    >
      <Portal
        container={
          <body
            theme="dark"
          />
        }
        id="connectionEditor"
        onClose={[Function]}
        onOpen={[Function]}
        onSelectTab={[Function]}
      />
    </Dialog>
    <Menu
      onItemClick={[Function]}
      orientation="column"
      placement="right-start"
    >
      <Popup
        className="msg menu"
        onClose={[Function]}
        onKeyDown={[Function]}
        onOpen={[Function]}
        orientation="column"
        pinned={false}
        placement="right-start"
        showArrow={false}
      >
        <Portal
          className="popupPortal"
          onClose={[Function]}
          onKeyDown={[Function]}
          onOpen={[Function]}
        />
      </Popup>
    </Menu>
  </ValueEditDialog>
  <ConfirmDialog
    id="confirmClearPasswordDlg"
  >
    <Dialog
      actions={
        {
          "end": [],
        }
      }
      caption={
        <g>
          <Icon
            disabled={false}
            src={278}
          />
          <Label>
            Confirm
          </Label>
        </g>
      }
      className="msg confirmDialog"
      container={
        <body
          theme="dark"
        />
      }
      content={
        <Container
          orientation="column"
        >
          
          <Container
            className="description"
            orientation="column"
          />
        </Container>
      }
      onClose={[Function]}
    >
      <Portal
        container={
          <body
            theme="dark"
          />
        }
        onClose={[Function]}
        onOpen={[Function]}
      />
    </Dialog>
  </ConfirmDialog>
  <ConfirmDialog
    id="confirmNewBastionDialog"
    onClose={[Function]}
  >
    <Dialog
      actions={
        {
          "end": [],
        }
      }
      caption={
        <g>
          <Icon
            disabled={false}
            src={278}
          />
          <Label>
            Confirm
          </Label>
        </g>
      }
      className="msg confirmDialog"
      container={
        <body
          theme="dark"
        />
      }
      content={
        <Container
          orientation="column"
        >
          
          <Container
            className="description"
            orientation="column"
          />
        </Container>
      }
      onClose={[Function]}
    >
      <Portal
        container={
          <body
            theme="dark"
          />
        }
        onClose={[Function]}
        onOpen={[Function]}
      />
    </Dialog>
  </ConfirmDialog>
</ConnectionEditor>
`;
